apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java

```diff
@@ -463,37 +496,79 @@ private String getOrDefault(JsonNode jsonNode) {
+                .flatMap(Collection::stream)
+                .map(entry -> Map.entry(entry.getKey(), JsonUtils.writeValueAsString(entry.getValue())))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
     private String getOrDefault(UUID value) {
         return Optional.ofNullable(value).map(UUID::toString).orElse("");
     }
 
     private Publisher<DatasetItem> mapItem(Result results) {
-        return results.map((row, rowMetadata) -> DatasetItem.builder()
-                .id(row.get("id", UUID.class))
-                .input(Optional.ofNullable(row.get("input", String.class))
-                        .filter(s -> !s.isBlank())
-                        .map(JsonUtils::getJsonNodeFromString).orElse(null))
-                .expectedOutput(Optional.ofNullable(row.get("expected_output", String.class))
-                        .filter(s -> !s.isBlank())
-                        .map(JsonUtils::getJsonNodeFromString).orElse(null))
-                .metadata(Optional.ofNullable(row.get("metadata", String.class))
-                        .filter(s -> !s.isBlank())
-                        .map(JsonUtils::getJsonNodeFromString).orElse(null))
-                .source(DatasetItemSource.fromString(row.get("source", String.class)))
-                .traceId(Optional.ofNullable(row.get("trace_id", String.class))
-                        .filter(s -> !s.isBlank())
-                        .map(UUID::fromString)
-                        .orElse(null))
-                .spanId(Optional.ofNullable(row.get("span_id", String.class))
-                        .filter(s -> !s.isBlank())
-                        .map(UUID::fromString)
-                        .orElse(null))
-                .experimentItems(getExperimentItems(row.get("experiment_items_array", List[].class)))
-                .lastUpdatedAt(row.get("last_updated_at", Instant.class))
-                .createdAt(row.get("created_at", Instant.class))
-                .createdBy(row.get("created_by", String.class))
-                .lastUpdatedBy(row.get("last_updated_by", String.class))
-                .build());
+        return results.map((row, rowMetadata) -> {
+            Map<String, DatasetItemInputValue<?>> inputData = Optional.ofNullable(row.get("input_data", Map.class))
+                    .filter(s -> !s.isEmpty())
+                    .map(value -> (Map<String, String>) value)
+                    .stream()
+                    .map(Map::entrySet)
+                    .flatMap(Collection::stream)
+                    .map(entry -> Map.entry(entry.getKey(),
+                            JsonUtils.readValue(entry.getValue(), new TypeReference<DatasetItemInputValue<?>>() {
+                            })))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            JsonNode input = getJsonNode(row, inputData, "input");
+            JsonNode expectedOutput = getJsonNode(row, inputData, "expected_output");
```

Comment:
Same, I'm not convinced about this resolution logic here. We need to think about the rollout for all the cases.

Reply:
Agreed to do this, in order to encapsulate the resolution logic in the BE.